Deprecate legacy services

joestringer opened the pull request:
Deprecate the `--enable-legacy-services` option and remove the old code.

Not all of the code is removed at this point, primarily just the bits that we should get in place during the v1.6 cycle so that we can fully remove the legacy code in v1.7. Most of it is removed, though (Some old interfaces and key/value types remain).

On the BPF side, as documented in `__ct_lookup()` and `ct_create4()`, in v1.6 we will populate the `ct_entry.rx_bytes` field with the backend id for the connection, and we will zero the `ct_entry.backend_id` (previously `ct_entry.slave`) field. When users upgrade to v1.6, the BPF code will start updating the conntrack entries such that the field will have no meaning; In v1.7, we can look at expanding the `ct_entry.backend_id` field to 32 bits (via a union) and retire the use of `ct_entry.rx_bytes` for LB conntrack entry backend id storage.

Related: https://github.com/cilium/cilium/issues/7967

Tasks:
* [x] Check it passes CI
* [x] Better PR description

In follow-up PRs:
* [ ] Look into removing the old interfaces, key/value types
* [ ] Documentation

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8419)
<!-- Reviewable:end -->

aanm reviewed (commented):
What is going to happen to the users coming from the 1.3 LTS? The default option is set to false in the configuration map.

borkmann reviewed (dismissed):
Given it's documented in the upgrade nodes, LGTM. At some point (perhaps 1.8) we can then finally remove all the old legacy LB code.

joestringer committed:
daemon: Deprecate `enable-legacy-services` option

Deprecate the option to enable legacy service map update. Retain the
environment binding so that Cilium will complain to users that they have
a deprecated configuration option enabled.

Retain the map cleanup, otherwise folks upgrading from an earlier
release with legacy services would end up with legacy BPF service maps
that are never cleaned up.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Remove testing of legacy services

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
examples: Remove legacy services option from CM

Remove the option from the ConfigMap, since it's disabled by default now
anyway.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
lbmap: Reuse serviceValueMap

Use serviceValueMap consistently across the code as the representation
of mappings from a backend (as a string) to its service value (initially
as a legacy service value, but soon as the service V2 value).

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
bpf: Remove legacy services

Continue to use the ct_entry.rx_bytes as the source of truth for the
service backend_id, as users upgrading from v1.5 may still have the
'slave' field being populated with the legacy services slave id.
Overwrite the 'slave' location in the ct_entry (now named 'backend_id')
with a zero so that subsequent releases can reuse it (eg, to exapnd the
backend_id space to 32 bits).

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
cilium: Remove legacy services dumping CLI

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
lbmap: Remove legacy service map manipulation

This commit removes the majority of the remaining legacy services code,
but leaves the map definitions in place along with some of the
interfaces, key/value types that are still used in a couple of places.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
lbmap: Store real BackendKey in cache

Store a real BackendKey (which is really just an interface around
loadbalancer.ID) in the lbmap cache, so that we don't have to have this
weird string hack to determine whether a cached backend is IPv6 or not.

Signed-off-by: Joe Stringer <joe@cilium.io>

brb commented on bpf/lib/conntrack.h:
```diff
@@ -230,14 +230,21 @@ static inline __u8 __inline__ __ct_lookup(void *map, struct __sk_buff *skb,
+			 * "ct_entry.rx_bytes" to store the services v2
+			 * backend (as it is not used with dir=CT_SERVICE).
+			 *
+			 * As of v1.6, "ct_entry.backend_id" is populated with
```
Unfortunately, we won't be easily able to store a backend ID in `ct_entry.backend_id`, because eventually, after https://github.com/cilium/cilium/pull/8041 has been merged, the backend ID will be `u32` , while `ct_entry.backend_id` is `u16`.

joestringer replied:
I see. Do you think it would help to zero the field instead? Then we could change the ct_entry to:
```
...
union {
    __u32 backend_id;
    struct {
        __u64 rx_bytes;
    };
}
...
```

joestringer replied:
^ This is what I've done in my latest push, let me know if you have other thoughts.

joestringer replied:
Sorry for the confusion, what I meant is that I changed it to zero the field.

I haven't modified the structure of ct_entry yet as I wasn't sure how it would impact the alignchecker.

brb reviewed (changes requested):
Thanks for working on this. I did my first round of review. LGTM besides the `ct_entry.backend_id` size issue. Will do another round to properly review the lbmap cache changes.

brb commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -494,24 +498,15 @@ func updateServiceV2Locked(fe ServiceKey, backends serviceValueMap,
 }
 
-func removeBackendsLocked(removedBackendIDs []loadbalancer.BackendID, isIPv6 bool,
+func removeBackendsLocked(removedBackendIDs []BackendKey, isIPv6 bool,
```
`isIPv6` is no longer needed. Can be addressed in a separate PR.

joestringer replied:
I need to rebase anyway, so I can follow up on this here.

brb commented on pkg/maps/lbmap/types.go:
```diff
@@ -32,12 +31,6 @@ import (
 
-// IsIPv6 detects in a dirty way whether the given backend addr ID belongs to
-// the ipv6 backend.
-func (b BackendAddrID) IsIPv6() bool {
```
:+1:

brb reviewed (approved):
Nice work! Also, thanks for making it reviewable per commit.

- Re https://github.com/cilium/cilium/pull/8419#discussion_r300055761: GH does not show such change. Maybe you forgot to push?
- Re removing legacy lbmap struct/interface definitions: did you kept them because of the map removal? Maybe we could introduce the `bpf.UnpinByPath()` function and get rid of them?
- Re remaining tasks: do you want to solve them in a separate PR?

Approved regardless of ^^ and one nit, as they can be resolved in a separate PR.

ianvernon reviewed (approved):
LGTM
